### PR TITLE
Refactor P3-AD interface to call only C++ version of P3

### PR DESCRIPTION
**Pull request by AaronDonahue**
This commit refactors the SCREAM-AD interface with P3 to use only the C++ implementation of P3.

This commit will also define a set of "universal" physics functions that can be accessed by any process.  These include, calculating Exner's formula given the pressure, converting temperature T into the potential temperature th, and the inverse, and calculating the width in meters for a specific level given the interface heights.

Addresses #759

### CODE REVIEW

**Comment by AaronDonahue** [comment]
[WIP] for now as this is still a work in progress.

**Comment by AaronDonahue** [comment]
TODO Item: Make unit tests for the new set of universal functions.

File: components/scream/src/physics/share/physics_functions.hpp
Comment (jeff-cohere) [inline components/scream/src/physics/share/physics_functions.hpp:102]: I think it would be nice to have a description of each of these "universal" functions with a bit more detail, including any assumptions about units, and the role of `range_mask`, since they could be used in various contexts.

Comment (AaronDonahue) [reply components/scream/src/physics/share/physics_functions.hpp:102]: Good point, Peter C suggested I break off the physics functions part into its own PR (see #805).  I'll make these changes there.

**Comment by AaronDonahue** [comment]
switching to WIP to save on AT cycles.
1. Need to determine why p3_stand_alone and p3_shoc_stand_alone tests are failing.
2. Waiting on #805 before merging anyways.

File: components/scream/src/physics/share/physics_functions.hpp
Comment (AaronDonahue) [reply components/scream/src/physics/share/physics_functions.hpp:102]: handles in #805

**Comment by AaronDonahue** [comment]
@jeff-cohere I've cleaned up the PR branch (note this is a different branch then the one I shared with you last week).  I'm going to wait for the AT before assigning reviewers.  But I don't anticipate any major deviations in this PR from the general structure here.

**Comment by AaronDonahue** [comment]
Ahhh, it looks like the p3_shoc coupled is failing, I know the issue.  I will have it fixed by tomorrow morning.

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (jeff-cohere) [inline components/scream/src/physics/p3/atmosphere_microphysics.cpp:46]: Maybe a silly question, but is there a "non-dimensional" unit supported by EKAT? "Meters per meter" sounds non-dimensional until I think about all the "kg / kg" and "mol / mol" stuff I've been seeing. @bartgol ?

Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:46]: not too silly.  This is kindof a hack, and maybe @bartgol  can comment on whether or not there is a legit nondim defined in EKAT.  kg/kg is a bit of a special unit because it represents kg/kg_air so it is "non-dimensional" but not unitless.  If that makes sense.

Comment (jeff-cohere) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:46]: It does. I'm actually in favor of more carefully annotating these quantities, at least in our documentation, to avoid confusing non-experts. In EAGLES we deal with various of those ratios involving "air", aerosols, and gases.

File: components/scream/src/physics/p3/atmosphere_microphysics.hpp
Comment (jeff-cohere) [inline components/scream/src/physics/p3/atmosphere_microphysics.hpp:87]: Do we explain what sort of "iteration" is going on anywhere? If not, can we add a little comment that does?

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (bartgol) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:46]: I thought I added a `ekat::units::nondim` declaration, but I guess I didn't. Doing `auto nondim = kg/kg` obviously work, and does the same thing. The mixing ratio is indeed unitless, except that we hack its string so that it prints `kg/kg`, by doing something like `auto q = kg/kg;  q.set_string("kg/kg");`. However, internally, `q` and `nondim` store the same "dimensions".

That said, we can declare a `nondim` unit in ekat. Or you could create one less ambiguously like so `ekat::Units nondim(0,0,0,0,0,0,0);`.

File: components/scream/tests/scream_p3/p3_init_vals.txt
Comment (jeff-cohere) [inline components/scream/tests/scream_p3/p3_init_vals.txt:1]: Is it possible to include a 1-line header that describes the names of the entities? Maybe not for this PR, but to make folks' lives easier in the future.

File: components/scream/tests/scream_p3_shoc/p3_init_vals.txt
Comment (jeff-cohere) [inline components/scream/tests/scream_p3_shoc/p3_init_vals.txt:1]: Same comment as for p3_init_vals.txt.

File: components/scream/src/physics/p3/p3_inputs_initializer.cpp
Comment (jeff-cohere) [inline components/scream/src/physics/p3/p3_inputs_initializer.cpp:96]: Just curious--is this not an error condition? It's true that no initialization has to be done in this case, but is it "valid" to call this method under such conditions?

**Review by jeff-cohere — APPROVED** [review]
This looks good, @AaronDonahue . I left a few questions/suggestions here and there, but I see nothing that should prevent this from going in. Nice work.

File: components/scream/src/physics/p3/p3_inputs_initializer.cpp
Comment (bartgol) [inline components/scream/src/physics/p3/p3_inputs_initializer.cpp:179]: You could avoid having `field_vals`, and just do 

```
int  icol, k, ipack, ivec;
s >> icol >> k;
ipack = k / n;
ivec = k % n;

s >> h_qv(icol,ipack)[ivec];
s >> h_th_atm(icol,ipack)[ivec];
s >> h_pmid(icol,ipack)[ivec];
...
```
`std::stringstream` is designed precisely to support this kind of (generic) input/output operations.

Note: if you need to skip some values, you can always do 
```
std::string skip;
s >> skip >> skip >> h_qv(icol,ipack)[ivec];
...
```

**Review by bartgol — CHANGES_REQUESTED** [review]
Great to see this getting close to be completed. I have a few comments/requests.

**Comment by AaronDonahue** [comment]
@bartgol  and @jeff-cohere , thank you for the comments and review.  I'm going to work on these now.

File: components/scream/src/physics/p3/CMakeLists.txt
Comment (AaronDonahue) [reply components/scream/src/physics/p3/CMakeLists.txt:15]: I'm not sure why this doesn't cause a link error.  Perhaps @jgfouca has an idea.  I'm pretty sure this is included here for the C++/F90 BFB unit tests.  I don't think any of the code in the AD interface call micro_p3.F90 directly.

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:5]: TBH I don't know.  My understanding is that it sets up the table lookup values used within p3_main.  @jgfouca , does that sound correct?

File: components/scream/src/physics/p3/CMakeLists.txt
Comment (jeff-cohere) [reply components/scream/src/physics/p3/CMakeLists.txt:15]: For what it's worth: if the function that needs `micro_p3_utils_init_c` isn't called itself, the linker is smart enough to eliminate the dependency.

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:46]: I think actually might have:
```  static Units nondimensional () {
    return Units(RationalConstant::one());
  }```

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:100]: We will need to store the two `_prev` fields in any restart file.  Given our current setup that would require these two fields to be registered with the field manager.

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:155]: Should the EKAT error message check be included in the `#ifndef`?

File: components/scream/src/physics/p3/p3_inputs_initializer.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/p3_inputs_initializer.cpp:96]: So I should switch this to EKAT_REQUIRE?

File: components/scream/src/physics/p3/p3_inputs_initializer.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/p3_inputs_initializer.cpp:179]: excellent, thank you for the suggestion.  I've cleaned this up and done this.

File: components/scream/src/physics/p3/p3_inputs_initializer.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/p3_inputs_initializer.cpp:167]: done

File: components/scream/tests/scream_p3/p3_init_vals.txt
Comment (AaronDonahue) [reply components/scream/tests/scream_p3/p3_init_vals.txt:1]: for sure, I made the change.

File: components/scream/src/physics/p3/p3_inputs_initializer.cpp
Comment (jeff-cohere) [reply components/scream/src/physics/p3/p3_inputs_initializer.cpp:96]: Sure, that works. We can always revisit if it turns out we're wrong about this.

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (bartgol) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:5]: It doesn't seem to write in any C++ provided ptr though (or in any array that is manually copied to a C++ ptr). My guess is that we don't need it. Have you tried to not call the f90 stuff and see what happens?

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (bartgol) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:101]: So is "T_prev" from the end of the last atm timestep? Or is it the end of the last time P3 ran (and if so, what if, say, P3 is subcycled)?

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (bartgol) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:157]: You mean the one at the bottom of the fcn? It should be inside the ifndef. Anything past `p3_init();` should be ifdef-ed out.

File: components/scream/src/physics/p3/CMakeLists.txt
Comment (bartgol) [reply components/scream/src/physics/p3/CMakeLists.txt:15]: Ah, it appears that our `physics_share` library compiles `eam/src/physics/cam/micro_p3_utils.f90`. I am a bit puzzled on why physics_share should depend on `micro_p3_utils.f90`. It seems to me that such file, as the name implies, is p3-specific.

@singhbalwinder digging in git history, it seems this dep was added when Murphy-Koop stuff was added. Do you know if this _really_ requires micro_p3_utils, or if it can be removed from `physics_share`?

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (jgfouca) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:5]: @bartgol , @AaronDonahue , I haven't thought about this in a while, but my recollection is we didn't want to bother converting initialization since we don't care about it's performance. p3_init must be called in order for various tables to be read and initialized.

Comment (bartgol) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:5]: TBC, I'm ok keeping initialization in F90. What puzzles me is that I can't see where the data is transferred from F90 to C. But if you know it is, than we're good.

File: components/scream/src/physics/p3/CMakeLists.txt
Comment (singhbalwinder) [reply components/scream/src/physics/p3/CMakeLists.txt:15]: I moved MurphyKoop in `wv_sat_scream.F90` so that both P3 and SHOC can use MurphyKoop. MurphyKoop depends on some variables which P3 initializes (zerodegc and ep_2), therefore I had that dependency. I agree that it should be removed. 

For removing this dependency, we might need to re-shuffle some codes and add codes to initialize these variables (zerodegc and ep_2) in the shared codes.

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (tcclevenger) [inline components/scream/src/physics/p3/atmosphere_microphysics.cpp:211]: Should your call to `auto elapsed_microsec = P3F::p3_main()` below come before you copy outputs back to device?

Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:211]: I believe that p3_main acts on device.  So we will need to make sure that host and device views are synced before calling p3_main.  Here I am removing the F90 version, which happened on host and replacing with the C++ version.  But, I may be misunderstanding how things work, it wouldn't be the first time, so I am open to some pushback.

File: components/scream/src/physics/p3/atmosphere_microphysics.hpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.hpp:181]: I removed m_it since it can now be stored in `infrastructure`, see line 190.  I also added a comment describing what it is even used for.  Question, would it make more sense to add that comment where the `P3Infrastructure` structure is defined?

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:101]: It should be from the last time `P3` is run.  But that is a good physics question, <- @hassanbeydoun or @PeterCaldwell .

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:202]: great, done.

**Comment by AaronDonahue** [comment]
@bartgol  and @jeff-cohere , I believe I have addressed your comments on this PR.  My changes were pretty robust so a few have disappeared from the code itself.  But I addressed the two biggest comments I think: a) Handling the structures passed to p3_main and b) How to deal with the local variables that are not stored in the field manager but are needed by P3. 

Can you take another look at the PR?  In particular, focus on the big changes I made to `atmosphere_microphysics.hpp`?

Last comment, while making these changes two big tasks occurred to me:
1. Still need a better unit test.
2. I have the pre-processing set up, but not post-processing.  Which will use the p3_main output to set variables needed by other processes.

The main question is:  Do we put this PR through now to get it out of the way and I begin work on these two tasks in a subsequent PR?  Or do we try to get these done here?  I'm partial to the former so that this PR doesn't get stale and it provides a good template for the SHOC stuff.

**Comment by jeff-cohere** [comment]
> The main question is: Do we put this PR through now to get it out of the way and I begin work on these two tasks in a subsequent PR? Or do we try to get these done here? I'm partial to the former so that this PR doesn't get stale and it provides a good template for the SHOC stuff.

I haven't looked over your latest changes yet, but for the record: I'm in favor of getting this PR merged, and creating issues to track the 2 outstanding items you mentioned.

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (bartgol) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:255]: Well, elapsed_microsec is not even used, so maybe we can remove it altogether?

File: components/scream/src/physics/p3/p3_inputs_initializer.cpp
Comment (jeff-cohere) [inline components/scream/src/physics/p3/p3_inputs_initializer.cpp:160]: Just a thought: using a file format that pairs field names with field data is more robust than assuming an ordering for the fields, since you can add and delete fields in the future without disrupting a sequence (and you can ignore any fields you don't need). Might not be worth fussing over, but I thought I'd mention it.

File: components/scream/src/physics/p3/atmosphere_microphysics.hpp
Comment (bartgol) [inline components/scream/src/physics/p3/atmosphere_microphysics.hpp:102]: I'm glad you enjoy so much the physics that you call it 'fun'. :)

File: components/scream/src/physics/p3/atmosphere_microphysics.hpp
Comment (jeff-cohere) [reply components/scream/src/physics/p3/atmosphere_microphysics.hpp:181]: Yes--if the iteration count is stored in a field for that struct, the documentation for it belongs there.

File: components/scream/src/physics/p3/atmosphere_microphysics.hpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.hpp:102]: yeah man, got make people excited to work with the `p3` code.  In a more serious response, it turns out that `physics` already has a defined meaning in `atmosphere_microphysics.hpp` so I had to add the fun.

File: components/scream/src/physics/p3/p3_inputs_initializer.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/p3_inputs_initializer.cpp:160]: I agree, in this case I was able to generate the input file directly from the `p3_run_and_cmp` output, which is why I kept it intact here, just in case there was some change in p3_run_and_cmp that I wanted reflected in the input file, and I didn't want to have to do any more formatting - i.e. lazy.

that being said, I didn't really put my heart into it because I think the more robust unit test will touch this code and would be the appropriate time to clean things up for good.

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:255]: ahhh I see, your comment is about `elapsed_microsec`.  It's true it isn't used, so @bartgol 's point is well-taken.  I'm not opposed to removing it, but that would require touching the unit testing code as well, which is maybe outside the scope of this PR (?)

Comment (bartgol) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:255]: I am not sure why it would touch the UT code. TBC, I'm not suggesting to make p3_main return void. I'm just suggesting we don't need the elapsed_microsec var here.

Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:255]: Is it possible to call a routine that returns `non-void` without assigning it some named output variable?

Comment (jeff-cohere) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:255]: Yep. Just omit the assignment.

Comment (AaronDonahue) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:255]: oh cool, well, that's been done.

**Review by bartgol — APPROVED** [review]
Let's merge this, as Jeff said, and fix the remaining issues later. Having this in can help the rest of the AD tasks to move fwd.

**Comment by AaronDonahue** [comment]
I forgot to add the "Automerge" label so I'm going to go ahead and merge this now.

**Pull request merged**
